src/codegen.pas

```diff
@@ -227,25 +229,26 @@ procedure RenderSongToGBDKC(Song: TSong; DescriptorName: String; Filename: strin
   OutSL.Add('');
 
   for I := Low(Song.Instruments.All) to High(Song.Instruments.All) do
-    with Song.Instruments.All[I] do begin
-      if SubpatternEnabled then begin
-        WriteStr(TypePrefix, Type_);
-        OutSL.Add(RenderGBDKSubpattern(TypePrefix+'SP' + IntToStr(ModInst(I)), Subpattern));
+    if InstrumentIsUsed(I, Song) then
```

Comment:
Isn't this going to break if, say, instruments 1 and 3 are used, but 2 isn't?

teNOR has to do [a whole song and dance of remapping instrument IDs](https://github.com/ISSOtm/fortISSimO/blob/bb12e7b4e93c2be54986e84233daaf059383df32/teNOR/src/optimise/mod.rs#L66-L67) when that happens, which requires some fairly complex copy-on-write semantics (because patterns, and especially instrument IDs, *can* be reused across channels that use *different* instrument "banks").

Reply:
That's good, but then I think that check should be outlined so that the logic appears sound without having to check out the function's definition. (I.e. what happened in my case—"wait, this doesn't sound right".)

A comment explaining *why* that extra condition is necessary would clear up any remaining confusion.